tests/test_site_apportionment/test_site_apportionment.py

```diff
@@ -634,3 +643,112 @@ def test_sort_rows_missing_cols(self, caplog):
+        "period"
+    ]
+    input_data = [
+        [1, 0, "0006", "RH12 1XL", 100.0, np.nan, "Clear", "R", "RH12 1XL", "202101"],
+        [1, 1, "0006", "RH12 1XL", 125.0, np.nan, "Clear", "R", "RH12 1XL", "202101"],
+        [1, 2, "0006", "RH12 1XL", np.nan, np.nan, "Clear", "R", "RH12 1XL", "202101"], # Nan 602 - Ensure conv to 0
+        [1, 2, "0006", "RH12 1XZ", 100.0, np.nan, "Clear", "R", "RH12 1XZ", "202101"], # different postcode
+        [2, 0, "0001", "NP44 2NZ", np.nan, 2.0, "Clear", "R", "NP44 2NZ", "202102"],
+        [2, 1, "0001", "NP44 2NZ", 50.0, 2.0, "Clear", "R", "NP44 2NZ", "202102"],
+        [3, 0, "0001", np.nan, np.nan, 1.0, "Check needed", "TMI", "NP30 7ZZ", "202102"], # NaN 601 - Ensure dropped
+    ]
+    input_df = pandasDF(data=input_data, columns=input_cols)
+
+    return input_df
+
+
+class TestCreateSitesDf(object):
+    """Tests for create_sites_df."""
+    
+    @pytest.fixture(autouse=True)
+    def set_attrs(self):
+        """Set class attributes that are passed as parameters."""
+        self.groupby_cols = ["reference", "period"]
+        self.site_cols = ["instance", "601", "602", "postcodes_harmonised"]
+    
+    def test_create_sites_df_raises(self, sites_df_input):
+        """Tests for create_sites_df when invalid data is passed."""
+        # test when a required column is missing
+        no_period = sites_df_input.copy()
+        no_period.drop("period", axis=1, inplace=True)
+        with pytest.raises(KeyError, match=r".*period.*not in index"):
+            create_sites_df(no_period, self.groupby_cols, self.site_cols)
+        # cols passed not in list[str] format
+        with pytest.raises(TypeError, match="unsupported operand type.*"):
+            create_sites_df(sites_df_input, 3, self.site_cols)
+
+    def test_create_sites_df_on_pass(self, sites_df_input):
+        """General tests for create_sites_df."""
+        output = create_sites_df(
+            sites_df_input, 
+            self.groupby_cols,
+            self.site_cols)
+        # assert the resultant dataframe is as expected
+        assert len(output) == 3, f"Output df has {len(output)} row. Expected 3"
+        expected_columns = ["reference",
+                            "period",
+                            "instance",
+                            "601",
+                            "602",
+                            "postcodes_harmonised"]
+        assert sorted(expected_columns) == sorted(output.columns.values)
+        assert np.array_equal(output["602"], [225.0, 100.0, 50.0]), (
+            "Column 602 (percent_col) has unexpected values."
+        )
+        assert np.array_equal(output["instance"], [0, 2, 0]), (
+            "Column 602 (percent_col) has unexpected values."
+        )
+        expected_postcodes = ["NP44 2NZ", "RH12 1XL", "RH12 1XZ"]
+        assert np.array_equal(
+            sorted(output["601"].unique()),
+            sorted(expected_postcodes)
+            ), "Postcodes not as expected"
```

Comment:
This approach is thorough. However, on this project our style has been to create the expected_output dataframe. I think this is in general a bit easier to read, as you can see at a glance what the shape and content of the expected output. This means the unit test also acts as an illustration of the action of the function. Would you be happy to update to this style?

Reply:
I could update it to that style, however I don't necessarily agree with it when you are looking to tests various parts and different inputs. Redefining input and output dfs many times over and over clutters the test file and leaves your test script as mainly just hard coded data.
In my opinion, asserting the output with corresponding failure message is sufficient, as the goal of the unit test is to ensure that the function is producing the expected output so that you know the functions work correctly.

Reply:
Hi Charlie, you make a good point, and I agree that the first aim of the unit test is to ensure each aspect of the test is covered. However, for now I think I'd prefer  us to follow what I take to be the style we use in the various RDSA projects I've worked on, which is to have hard coded input and expected output dataframes. 
I think the advantage of this is that it allows people with limited coding experience to see at a glance what the function does.

Let's take the debate on different ways of doing unit tests to an RDSA best practice chat some time :-)